flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -4394,11 +4450,54 @@ class ArrayExprLowering {
         }(),
         x);
   }
+  static mlir::Type unwrapBoxEleTy(mlir::Type ty) {
+    if (auto boxTy = ty.dyn_cast<fir::BoxType>()) {
+      ty = boxTy.getEleTy();
+      if (auto refTy = fir::dyn_cast_ptrEleTy(ty))
+        ty = refTy;
+    }
+    return ty;
```

Comment:
Wouldn't `fir::dyn_cast_ptrOrBoxEleTy()` yield the same result on the intended use cases of `unwrapBoxEleTy` (fir.box or non memory `ty`) ?